### PR TITLE
Adds support for label blocks

### LINKED ISSUE #408: Labels don't handle indentation properly
Stuff like this causes an indentation error:
```
label_name:
    some_code()
    if(prob(50))
        goto label_name
```

### PR DESCRIPTION
**Pull request by ike709**
Fixes #408 

Adds support for label blocks and using continue/break in them.

This survived extremely basic testing but could probably use more rigorous BYOND parity testing.

### CODE REVIEW

File: DMCompiler/DM/DMProc.cs [hunk 1]
```diff
@@ -227,16 +227,30 @@ private class DMProcScope {
...
         public void BreakIfFalse() {
             JumpIfFalse(_loopStack.Peek() + "_end");
         }
 
-        public void Continue() {
-            Jump(_loopStack.Peek() + "_continue");
+        public void Continue(DMASTIdentifier label = null) {
+            if (label is not null)
+            {
+                Jump(label.Identifier + "_start");
```
Comment (wixoaGit) [inline DMCompiler/DM/DMProc.cs:248]: This doesn't continue a `for` loop properly. A `continue` needs to jump to a for's incrementor, which is at the end of the loop just before it jumps to the beginning.

A quirk to take note of is that label bodies can contain more than just a loop, and `continue label` will always continue the first loop in the body. `break label` will always exit the label body entirely.
```
var/i = 1
a:
	world.log << "A"
	for (,TRUE,i++)
		world.log << i
		break a
	world.log << "B"
sleep(10)
continue a
```
Outputs:
```
A
1
2
3
...
```

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 2]
```diff
@@ -591,7 +591,14 @@ public DMASTProcStatement ProcStatement()
...
-            if (Check(TokenType.DM_Break)) {
+            if (Check(TokenType.DM_Break))
+            {
+                Whitespace();
+                DMASTExpression expression = Expression();
+                if (expression is DMASTIdentifier)
+                {
+                    return new DMASTProcStatementBreak((DMASTIdentifier)expression);
+                }
                 return new DMASTProcStatementBreak();
```
Comment (wixoaGit) [inline OpenDreamShared/Compiler/DM/DMParser.cs:602]:
```suggestion
                DMASTExpression label = Expression();
                
                return new DMASTProcStatementBreak(label as DMASTIdentifier);
```

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 3]
```diff
@@ -600,6 +607,12 @@ public DMASTProcStatement ProcStatement()
 
         public DMASTProcStatementContinue Continue() {
             if (Check(TokenType.DM_Continue)) {
+                Whitespace();
+                DMASTExpression expression = Expression();
+                if (expression is DMASTIdentifier)
+                {
+                    return new DMASTProcStatementContinue((DMASTIdentifier)expression);
+                }
                 return new DMASTProcStatementContinue();
```
Comment (wixoaGit) [inline OpenDreamShared/Compiler/DM/DMParser.cs:616]:
```suggestion
                DMASTExpression label = Expression();
                
                return new DMASTProcStatementContinue(label as DMASTIdentifier);
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 4]
```diff
@@ -1148,6 +1155,21 @@ public DMASTProcStatementThrow Throw()
...
+        {
+            Whitespace();
+            Newline();
+
+            DMASTProcBlockInner body = ProcBlock();
+            if (body == null) {
+                DMASTProcStatement statement = ProcStatement();
+                
+                if (statement == null) return new DMASTProcStatementLabel(expression.Identifier, null);
+                body = new DMASTProcBlockInner(new DMASTProcStatement[] { statement });
```
Comment (wixoaGit) [inline OpenDreamShared/Compiler/DM/DMParser.cs:1168]:
```suggestion
                if (statement != null) body = new DMASTProcBlockInner(new DMASTProcStatement[] { statement });
```

File: DMCompiler/DM/Visitors/DMProcBuilder.cs [hunk 5]
```diff
@@ -111,10 +111,20 @@ class DMProcBuilder {
 
         public void ProcessStatementLabel(DMASTProcStatementLabel statementLabel) {
             _proc.AddLabel(statementLabel.Name + "_codelabel");
+            if (statementLabel.Body is not null)
+            {
+                _proc.StartScope();
+                {
+                    _proc.AddLabel(statementLabel.Name + "_start");
```
Comment (wixoaGit) [inline DMCompiler/DM/Visitors/DMProcBuilder.cs:118]: This label is not used at all
```suggestion
```

File: DMCompiler/DM/DMProc.cs [hunk 6]
```diff
@@ -245,16 +246,46 @@ private class DMProcScope {
...
+            if (label is not null)
+            {
+                var codeLabel = label.Identifier + "_codelabel";
+                if (!_labels.ContainsKey(codeLabel))
+                {
+                    Program.Error(new CompilerError(null, $"Unknown label {label.Identifier}"));
+                }
+                var labelList = _labels.Keys.ToList();
+                var continueLabel = string.Empty;
+                for (var i = labelList.IndexOf(codeLabel) + 2; i < labelList.Count; i++)
```
Comment (wixoaGit) [inline DMCompiler/DM/DMProc.cs:275]: For the previous suggestion
```suggestion
                for (var i = labelList.IndexOf(codeLabel) + 1; i < labelList.Count; i++)
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Comment by ike709** [comment]
Reviews addressed

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**